File: c++/Board.cxx

#include "./Board.hxx"

bool Board::check(const Guess& guess) const
{
    int hit = 0;
    int colorOnlyHit = 0;
    int usedSecrets = 0;
    for (int i = 0 ; i < 4 ; i++)
    {
        if (guess.get(i) == secret->get(i))
        {
            hit++;
            // cout << "Guess " << i << " is HIT \r\n";
        }
        else
        {
            bool found = false;
            for (int j = 0 ; j < 4 ; j++)
            {
                if (i != j && secret->get(j) == guess.get(i) && !(secret->get(j) == guess.get(j)) && !(usedSecrets & (1<<j)))
                {
                    colorOnlyHit++;
                    usedSecrets |= (1<<j);
                    // cout << "Guess " << i << " found a matching color elsewhere \r\n";
                    found = true;
                    break;
                }
            }
            if (!found) {
                // cout << "Guess " << i << " missed. \r\n";
            }
        }
    }
    if (hit == 4) 
    {
        return true;
    }
    cout << "You guessed " << hit << " pins correctly and found " << colorOnlyHit << " matching colors elsewhere. \r\n";
    return false;
}

const Secret* Board::getSecret() const
{
    return this->secret;
}

File: swift/Sources/MasterMind/Board.swift
//
//  Board.swift
//  MasterMind
//
//  Created by oglimmer on 11/12/21.
//

import Foundation

public class Board : CustomStringConvertible{

    var secret: Secret;

    init() {
        secret = Secret()
    }

    init(secret: Secret) {
        self.secret = secret;
    }

    func check(guess: Guess) -> CheckResult {
        let checkResult: CheckResult = CheckResult()
        var bitSet: UInt8 = 0
        for i in 0...3 {
            if guess.get(pos: i).identifier == secret.get(pos: i).identifier {
                checkResult.incHit()
            } else {
                for j in 0...3 {
                    if secret.get(pos: j).identifier == guess.get(pos: i).identifier &&
                            secret.get(pos: j).identifier != guess.get(pos: j).identifier &&
                            ((bitSet & (1<<j)) == 0) {
                        checkResult.incColorOnlyHit()
                        bitSet |= (1<<j)
                        break
                    }
                }
            }
        }
        return checkResult
    }

    func checkAndInform(guess: Guess) -> CheckResult {
        let checkResult: CheckResult = check(guess: guess)
        if !checkResult.hasWon() {
            print("You guessed " + String(checkResult.getHit()) + " pins correctly and found " + String(checkResult.getColorOnlyHit()) + " matching colors elsewhere.")
        }
        return checkResult
    }

    public var description: String {
        return secret.description
    }

}

